consensus/safety-rules/src/safety_rules.rs

```diff
@@ -303,9 +266,42 @@ impl SafetyRules {
 
             info!(SafetyLogSchema::new(LogEntry::Epoch, LogEvent::Update).epoch(epoch_state.epoch));
         }
-        self.epoch_state = Some(epoch_state);
+        self.epoch_state = Some(epoch_state.clone());
 
-        Ok(())
+        let author = self.persistent_storage.author()?;
+        let expected_key = epoch_state.verifier.get_public_key(&author);
+        let initialize_result = match expected_key {
+            None => Err(Error::ValidatorNotInSet(author.to_string())),
+            Some(expected_key) => {
+                let current_key = self.signer().ok().map(|s| s.public_key());
+                if current_key == Some(expected_key.clone()) {
+                    debug!(
+                        SafetyLogSchema::new(LogEntry::KeyReconciliation, LogEvent::Success),
+                        "in set",
+                    );
+                    Ok(())
+                } else {
+                    match self
+                        .persistent_storage
+                        .consensus_key_for_version(expected_key)
+                    {
+                        Err(error) => Err(Error::ValidatorKeyNotFound(error.to_string())),
```

Comment:
Though we slightly mask it here.

Reply:
True.. although this is exactly what we do today -- we catch the error and say "Validator key not found: {error}"